tests/unittests/Foundation/NSUndoManagerTests.mm

```diff
@@ -0,0 +1,718 @@
+//******************************************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <Foundation/Foundation.h>
+#import <TestFramework.h>
+
+@interface UndoTestObject : NSObject
+@property (retain) NSUndoManager* undoManager;
+@property (retain) NSNumber* value;
+@property (retain) NSMutableArray* arrayOfNumbers;
+- (id)initWithValue:(NSNumber*)num;
+- (id)initWithEmptyArray;
+- (void)updateNumber:(NSNumber*)value;
+- (void)updateNumberSingleUndoStep:(NSNumber*)value;
+- (void)addToMutableArray:(NSNumber*)value;
+- (void)removeFromMutableArray:(NSNumber*)value;
+@end
+
+@implementation UndoTestObject
+
+- (id)initWithValue:(NSNumber*)num {
+    self = [super init];
+    _undoManager = [[NSUndoManager alloc] init];
+    _value = [num retain];
+    _arrayOfNumbers = nil;
+    return self;
+}
+
+- (id)initWithEmptyArray {
+    self = [super init];
+    _undoManager = [[NSUndoManager alloc] init];
+    _value = nil;
+    _arrayOfNumbers = [[NSMutableArray array] retain];
+    return self;
+}
+
+- (void)dealloc {
+    [_undoManager release];
+    [_value release];
+    [_arrayOfNumbers release];
+}
+
+- (void)updateNumber:(NSNumber*)newValue {
+    [self.undoManager registerUndoWithTarget:self selector:@selector(updateNumber:) object:self.value];
+    self.value = newValue;
+}
+
+- (void)updateNumberSingleUndoStep:(NSNumber*)value {
+    [self.undoManager beginUndoGrouping];
+    [self updateNumber:value];
+    [self.undoManager endUndoGrouping];
+}
+
+- (void)addToMutableArray:(NSNumber*)value {
+    [self.undoManager registerUndoWithTarget:self selector:@selector(removeFromMutableArray:) object:value];
+    [self.arrayOfNumbers addObject:value];
+}
+
+- (void)removeFromMutableArray:(NSNumber*)value {
+    [self.undoManager registerUndoWithTarget:self selector:@selector(addToMutableArray:) object:value];
+    [self.arrayOfNumbers removeObject:value];
+}
+
+@end
+
+TEST(NSUndoManager, BasicUndoTest) {
+    NSNumber* oldValue = [NSNumber numberWithInt:5];
+    NSNumber* newValue = [NSNumber numberWithInt:6];
+
+    UndoTestObject* object = [[[UndoTestObject alloc] initWithValue:oldValue] autorelease];
+    EXPECT_EQ([object.undoManager canUndo], NO);
+
+    EXPECT_EQ(object.value.intValue, oldValue.intValue);
```

Comment:
>EXPECT_EQ [](start = 4, length = 9)

nit: You can compare NSNumber instances with `OBJCEQ` instead of querying their primitive values...